続きのレッスン取得機能のためにchapterIdのレスポンスを返すよう変更【チャプター&レッスン一覧画面】※受講生側（DAIKI）

i-k-i-a-d opened the pull request:
やったこと
未完了かつ、Lessonidが一番小さいLessonidと一緒に、それを含むチャプターidも返せるように実装。
progressメソッドのResourceクラスの戻り値を微修正。

確認方法
Postmanで受講生1のデータでログインし、以下URLをSend。
http://localhost:8080/api/v1/attendance/1/progress

実行結果
{
    "data": {
        "attendance": {
            "attendance_id": 1,
            "progress": 10,
            "course": {
                "course_id": 1,
                "title": "PHP入門講座",
                "image": "course/4459908b-3cdf-4521-94fa-c2a9746d92e1.png"
            }
        },
        "number_of_completed_chapters": 1,
        "number_of_total_chapters": 3,
        "number_of_completed_lessons": 2,
        "number_of_total_lessons": 6,
        "continue_lesson": {
            "chapter_id": 2,
            "lesson_id": 3
        }
    }
}

i-k-i-a-d committed:
未完了かつ、Lessonidが
一番小さいLessonidと一緒に、それを含むチャプターidも返せるように実装。
progressメソッドのResourceクラスの戻り値を微修正。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -112,22 +112,23 @@ private function getTotalLessonsCount($attendance)
         $youngestUnCompletedLessonId = null;
-        $attendance->course->chapters->each(function ($chapter) use ($attendance, &$youngestUnCompletedLessonId) {
-            $chapter->lessons->each(function ($lesson) use ($attendance, &$youngestUnCompletedLessonId) {
+        $youngestUnCompletedChapterId = null;
```
ここは同一変数として定義した方が良いと思います。

````php
        $youngestUnCompletedLesson = [
            'chapter_id' => null,
            'lesson_id' => null,
        ];
````

i-k-i-a-d replied:
言われてみればそうですね。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -112,22 +112,23 @@ private function getTotalLessonsCount($attendance)
      *
      * @param Attendance $attendance
-     * @return int|null
+     * @return array $continueLesson
      */
     private function getYoungestUnCompletedLessonId($attendance)
     {
```
phpdocのreturnの記述はデータ型が分かれば問題ないです！
この関数を呼び出すときに返り値の変数名は知る必要がないので！

````php
@return array | null
````

kou-tech replied:
関数名もメンテナンスしましょう！

`getYoungestUnCompletedLesson`だと適切だと思います！

i-k-i-a-d replied:
前に自分で調べた記述に変数まで記述があったのでマネしてました。

kou-tech replied:
なるほどです！現場によっては、関数のドキュメントとして変数を書く場合もあるかもしれないです！
私はよく静的解析のための型づけのために利用していたので、使用用途が違うかもしれないですね。

https://tech-blog.rakus.co.jp/entry/20210326/php#PHP%E3%81%AE%E5%9E%8B%E3%82%92%E4%BD%BF%E3%81%84%E3%81%93%E3%81%AA%E3%81%97%E3%81%9F%E3%81%84

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -136,7 +137,9 @@ private function getYoungestUnCompletedLessonId($attendance)
             });
         });
-        return $youngestUnCompletedLessonId;
+        return $continueLesson = [
+            'chapter_id' => $youngestUnCompletedChapterId,
+            'lesson_id' => $youngestUnCompletedLessonId,
+        ];
```
変数名にわざわざ格納する必要はないです！

````php
        return [
            'chapter_id' => $youngestUnCompletedChapterId,
            'lesson_id' => $youngestUnCompletedLessonId,
        ];
````

kou-tech replied:
そして、具体的な話になるのですが、`$youngestUnCompletedLesson`の`lesson_id`がnullの場合はこの関数はnullを返すようにして欲しいです！

i-k-i-a-d replied:
確かに、回りくどい記述でしたね。
まとめてnullをかえすように実装してます。

kou-tech commented on app/Http/Resources/Student/AttendanceCourseProgressResource.php:
```diff
@@ -31,7 +30,7 @@ public function toArray($request)
             "number_of_total_chapters" => $progressData['totalChaptersCount'],
             "number_of_completed_lessons" => $progressData['completedLessonsCount'],
             "number_of_total_lessons" => $progressData['totalLessonsCount'],
-            "continue_lesson_id" => $progressData['youngestUnCompletedLessonId'],
+            "continue_lesson" => $progressData['youngestUnCompletedLessonId'],
         ];
```
`$progressData['youngestUnCompletedLessonId']`もIdのみではなくなったので、違う変数名に修正をお願いします！

i-k-i-a-d replied:
ここも変えるべきでしたね。

i-k-i-a-d committed:
別々の変数名でnullを定義していたが、配列でまとめて定義するよう修正。
phpDocの戻り値の記述を型だけに修正。
メソッド名を修正。
メソッドの戻り値の記述を修正。
メソッドの戻り値でlesson_idの値がnullだったら戻り値もnullで返すように実装。
progressメソッド戻り値の配列のキーを修正。
eachメソッドのループで値を取得した後も、ループが続かないように条件分岐を実装。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -112,31 +112,36 @@ private function getTotalLessonsCount($attendance)
                 if ($lessonAttendance->status !== LessonAttendance::STATUS_COMPLETED_ATTENDANCE) {
-                    if ($youngestUnCompletedLessonId === null) {
-                        $youngestUnCompletedLessonId = $lesson->id;
+                    if ($youngestUnCompletedLesson['lesson_id'] === null && $youngestUnCompletedLesson['chapter_id'] === null) {
+                        $youngestUnCompletedLesson['lesson_id'] = $lesson->id;
+                        $youngestUnCompletedLesson['chapter_id'] = $chapter->id;
```
条件は以下だけで問題ないです！
というのも、`$youngestUnCompletedLesson['chapter_id'] === null`このケースは想定されないからです。

````php
$youngestUnCompletedLesson['lesson_id'] === null
````

kou-tech replied:
これは好みですが、同時に配列の要素を更新していることがわかるように次のように書くと良いかもしれないです。

````php
                        $youngestUnCompletedLesson = [
                            'lesson_id' => $lesson->id,
                            'chapter_id' => $chapter->id,
                        ];
````

i-k-i-a-d replied:
確かに、未完了のlesson_idが取れてないときは、それを含むチャプターidもnullのままなので、二つも記述する必要ないですね。

また、配列で取得するように記述した方が見やすいですね。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -112,31 +112,36 @@ private function getTotalLessonsCount($attendance)
-                    }
                 }
             });
+            if ($youngestUnCompletedLesson['lesson_id'] !== null && $youngestUnCompletedLesson['chapter_id'] !== null) {
+                return;
+            }
```
これも次の条件式だけで問題ないです！

````php
$youngestUnCompletedLesson['lesson_id'] !== null
````

i-k-i-a-d replied:
修正しました。

ここの条件式は自分が記述するまで実装されてなかったのですが、これがないと、配列の値を取得した後も無駄にeachメソッドのループが続くことになるので、あった方がいいですよね？

kou-tech replied:
それであれば次の記述の方が適切だと思います！
というのも、eachはbreakできないので早期returnで次の要素へcontinueする方法を取ります。

````php
        $attendance->course->chapters->each(function ($chapter) use ($attendance, &$youngestUnCompletedLesson) {
            if ($youngestUnCompletedLesson['lesson_id'] !== null) {
                return;
            }

            $chapter->lessons->each(function ($lesson) use ($attendance, &$youngestUnCompletedLesson, $chapter) {
                $lessonAttendance = $attendance->lessonAttendances->where('lesson_id', $lesson->id)->first();
                if ($lessonAttendance->status !== LessonAttendance::STATUS_COMPLETED_ATTENDANCE) {
                    if ($youngestUnCompletedLesson['lesson_id'] === null) {
                        $youngestUnCompletedLesson = [
                            'lesson_id' => $lesson->id,
                            'chapter_id' => $chapter->id,
                        ];
                        return;
                    }
                }
            });
        });
````

i-k-i-a-d committed:
未完了のlesson_idとそれを含むチャプターのidを取得するメソッドの値を取得する記述と条件分岐を修正。

Commit:
Fix incomplete lesson check in
AttendanceController
